### DISCUSSION TITLE
`ref` parameters on Primary constructors

### DISCUSSION THREAD

**Discussion by FaustVX** [dotnet/csharplang#7343] [General] [answered]
Hi,

I just tried a few thing with primary constructors.
And I found a little bug in the compiler (I think).

You can't assign a `ref` parameter to a `ref` field/property in a primary ctor,

```C#
readonly ref struct C1(ref object ptr)
{
    public readonly ref object _ptr = ref ptr; // CS9077: Cannot return a parameter by reference 'ptr' through a ref parameter; it can only be returned in a return statement
}
```

But it works well in other scenario:
- `ref` of other field in Primary ctor
  ```C#
  readonly ref struct C1(ref object ptr)
  {
      public static object obj = new();
      public readonly ref object _ptrObj = ref obj; // Ok
  }
  ```
- `ref` of `in` parameter in Primary ctor (using `Unsafe.AsRef()`)
  ```C#
  readonly ref struct C2(in object ptr)
  {
      public readonly ref object _ptr = ref Unsafe.AsRef(in ptr); // Ok
  }
  ```

But I found a workaround, using `Unsafe.AsRef()` of a `ref` parameter
```C#
readonly ref struct C3(ref object ptr)
{
    public readonly ref object _ptr = ref Unsafe.AsRef(in ptr); // Workaround
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwgBuE2YBTAJiAGoAPgAEADAAIRARgB0AJThkYuCLNI4ADlgA2nAMqc2WGigDcAWABQ1pFF7kdBCUjATK1OhKJjrAb2sJIKkAZilpADYJEmAAKwgvGNiJAF4JMggAdwAKAEpLK2DQl2gHMicStySEmAkAfU0qAHk41MrouLMJAHpuiSaAa0DgkTC7MorXDvivBqoCotHvMWyp6q9GhFzhoqCUgD56zbapza7e/qGrAF9rW1LHZymPWlqiaVWIKriaiU3tqwBQojMIyKLrWpJNoZHL5HbFcaPdoQo7NVppNadHp9QbwpaI8pPL7TX5zBAnYlnbHefQATjEAHYGSBvFAyGQSLURAyJFA/vgoDgIDBOBJgETIEgyKYJAByTayiQwAAWCBIcAA5sree1NAKhSKEF0sLUaGzooSxRApAyILwJFgyDrue4YLAIELlNZbjYrATJsSXl4iAAmbKOkkbKgAoGLUGRSOQ9HpLJ5BYgkr2JGYma1MktZIY4lJc44q5xzMTCWJ1HkotuACqFCgkFkAEFkPIvuGnf9S5dvXc/Q9Lc8qK9vCFPt9c39o/48WMRwGZ6TjvWJE3kC21B2u2Ae3Otv2AOoMAb4dVkXjeoA)

**Comment by 333fred** [comment]
Please file a bug on roslyn. @AlekseyTs fyi.

**Reply by FaustVX** [reply, marked as answer]
Ok, opened dotnet/roslyn#69055
